AX: AccessibilityScrollView::parentObject should try harder to return a valid parent AccessibilityObject

minorninth opened the pull request:
#### a3a366646035270931774ab6e4fe35555626766e
<pre>
AX: AccessibilityScrollView::parentObject should try harder to return a valid parent AccessibilityObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=296205">https://bugs.webkit.org/show_bug.cgi?id=296205</a>
<a href="https://rdar.apple.com/156171319">rdar://156171319</a>

Reviewed by Tyler Wilcock.

We determined the root cause of a missing content bug was mistakenly
interpreting a AccessibilityScrollView from an iframe as the root
AccessibilityScrollView for the whole page.

While the primary fix is to use more robust logic before updating the
tree root, it also makes sense for AccessibilityScrollView to be more
robust in returning a parent, even if its owner element doesn&apos;t happen
to return a renderer. It&apos;s probably better to keep walking up the node
hierarchy rather than returning null.

* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::parentObject const):

Canonical link: <a href="https://commits.webkit.org/297651@main">https://commits.webkit.org/297651@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0cbacc333fd5c859091723f0dc71dda106b79ad1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112389 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32121 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22598 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118468 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62772 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114351 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32773 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40684 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85379 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/36085 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115336 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26155 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101116 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65810 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25450 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19254 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62313 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95539 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19329 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121794 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39463 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29381 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94185 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39844 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97356 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94010 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24043 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39255 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17054 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35513 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39351 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44839 "") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38986 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42323 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40729 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AccessibilityScrollView.cpp:
```diff
@@ -338,15 +338,20 @@ AccessibilityObject* AccessibilityScrollView::parentObject() const
-    return nullptr;
+        parentElement = remoteFrameView->frame().ownerElement();
+
+    AccessibilityObject* parentAccessibilityObject = nullptr;
```
Maybe `ancestorAccessibilityObject`, since we may end up traversing past our direct parent? And `ancestorElement` over `parentElement` too?

twilco commented on Source/WebCore/accessibility/AccessibilityScrollView.cpp:
```diff
@@ -338,15 +338,20 @@ AccessibilityObject* AccessibilityScrollView::parentObject() const
+    while (parentElement && !parentAccessibilityObject) {
+        parentAccessibilityObject = cache->getOrCreate(*parentElement);
+        if (parentAccessibilityObject)
+            break;
```
Feel free to not do with this if you think the way you've written is cleaner. But a way to do this in one line would be:

```cpp
if ((ancestorAccessibilityObject = cache->getOrCreate(*ancestorElement))
    break;
```

minorninth committed:
AX: AccessibilityScrollView::parentObject should try harder to return a valid parent AccessibilityObject
https://bugs.webkit.org/show_bug.cgi?id=296205
rdar://156171319

Reviewed by Tyler Wilcock.

We determined the root cause of a missing content bug was mistakenly
interpreting a AccessibilityScrollView from an iframe as the root
AccessibilityScrollView for the whole page.

While the primary fix is to use more robust logic before updating the
tree root, it also makes sense for AccessibilityScrollView to be more
robust in returning a parent, even if its owner element doesn't happen
to return a renderer. It's probably better to keep walking up the node
hierarchy rather than returning null.

* Source/WebCore/accessibility/AccessibilityScrollView.cpp:
(WebCore::AccessibilityScrollView::parentObject const):

Canonical link: https://commits.webkit.org/297651@main
